pkg/controllers/lballoc/lballoc.go

```diff
@@ -0,0 +1,517 @@
+package lballoc
+
+import (
+	"context"
+	"errors"
+	"net"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/cloudnativelabs/kube-router/v2/pkg/healthcheck"
+	"github.com/cloudnativelabs/kube-router/v2/pkg/options"
+	v1core "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+)
+
+const loadBalancerClassName = "kube-router"
+
+type ipRanges struct {
+	ipRanges   []net.IPNet
+	rangeIndex int
+	currentIP  net.IP
+}
+
+type LoadBalancerController struct {
+	ipv4Ranges   *ipRanges
+	ipv6Ranges   *ipRanges
+	svcLister    cache.Indexer
+	lock         *resourcelock.LeaseLock
+	addChan      chan v1core.Service
+	allocateChan chan v1core.Service
+	clientset    kubernetes.Interface
+	isDefault    bool
+	syncPeriod   time.Duration
+}
+
+func getNamespace() (namespace string, err error) {
+	ns := os.Getenv("POD_NAMESPACE")
+	if ns != "" {
+		return ns, nil
+	}
+
+	nb, err := os.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+	if err == nil {
+		return string(nb), nil
+	}
+
+	return "", errors.New("unable to get namespace from kubernetes environment or $POD_NAMESPACE")
+}
+
+func getPodname() (podname string, err error) {
+	podname = os.Getenv("POD_NAME")
+	if podname != "" {
+		return podname, nil
+	}
+	return "", errors.New("unable to get pod name from $POD_NAME")
+}
+
+func copyIP(ip net.IP) net.IP {
+	return append(net.IP{}, ip...)
+}
+
+func newipRanges(ranges []net.IPNet) *ipRanges {
+	var cip net.IP
+	if len(ranges) == 0 {
+		cip = nil
+	} else {
+		cip = copyIP(ranges[0].IP)
+	}
+	ir := &ipRanges{
+		ipRanges:  ranges,
+		currentIP: cip,
+	}
+
+	return ir
+}
+
+func (ir *ipRanges) inc() {
+	cn := ir.ipRanges[ir.rangeIndex]
+	ci := copyIP(ir.currentIP)
+
+	// Increment the current IP address
+	// 10.0.0.3 will increment to 10.0.0.4
+	// 10.0.0.255 will increment to 10.0.1.0
+	for i := len(ci) - 1; i >= 0; i-- {
+		ci[i]++
+		if ci[i] > 0 { // if the byte didn't overflow to zero, don't increment the byte to the left
+			break
+		}
+	}
+
+	// If the new address is not in the current IP range, move to the first IP in the next range
+	// If the current range is the last, move to the first IP in the first range
+	if !cn.Contains(ci) {
+		if ir.rangeIndex == len(ir.ipRanges)-1 {
+			ir.rangeIndex = 0
+		} else {
+			ir.rangeIndex++
+		}
+		ci = copyIP(ir.ipRanges[ir.rangeIndex].IP)
+	}
+
+	ir.currentIP = ci
+}
+
+func ipInAllocated(ip net.IP, allocated []net.IP) bool {
+	for _, cip := range allocated {
+		if cip.Equal(ip) {
+			return true
+		}
+	}
+	return false
+}
+
+func (ir *ipRanges) getNextFreeIP(allocated []net.IP) (net.IP, error) {
+	startIP := copyIP(ir.currentIP)
+	if len(startIP) == 0 {
+		return nil, errors.New("no IPs left to allocate")
+	}
+	ip := startIP
+	for {
+		if !ipInAllocated(ip, allocated) {
```

Comment:
You probably need to ensure that it is `<255` as well. I didn't get that high in my tests, but 255 is typically the broadcast address and is also not assignable.

Reply:
That's a good point. I guess I'm so used to using IPs within network, somehow using `.0` just feels so gross to me. But not only for small ranges, but also larger ranges, unless you're going to parse the CIDR for the network and broadcast address, it means that you would be losing 2 IPs for ever /24 that your CIDR contains.

It might be worth mentioning this in the doc as well, but I could also just be nervous over my own handling of `.0` and `.255` :rofl: